File: IAIso-v5.0/core/iaiso-java/iaiso-conformance/src/main/java/io/iaiso/conformance/ConsentRunner.java
package io.iaiso.conformance;

import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import io.iaiso.consent.*;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;

final class ConsentRunner {
    private ConsentRunner() {}

    static List<VectorResult> run(Path specRoot) throws IOException {
        Path file = specRoot.resolve("consent").resolve("vectors.json");
        JsonObject doc = JsonParser.parseString(
            new String(Files.readAllBytes(file))).getAsJsonObject();
        String sharedKey = doc.get("hs256_key_shared").getAsString();

        List<VectorResult> out = new ArrayList<>();
        if (doc.has("scope_match")) {
            for (JsonElement el : doc.getAsJsonArray("scope_match")) {
                out.add(runScopeMatch(el.getAsJsonObject()));
            }
        }
        if (doc.has("scope_match_errors")) {
            for (JsonElement el : doc.getAsJsonArray("scope_match_errors")) {
                out.add(runScopeMatchError(el.getAsJsonObject()));
            }
        }
        if (doc.has("valid_tokens")) {
            for (JsonElement el : doc.getAsJsonArray("valid_tokens")) {
                out.add(runValidToken(sharedKey, el.getAsJsonObject()));
            }
        }
        if (doc.has("invalid_tokens")) {
            for (JsonElement el : doc.getAsJsonArray("invalid_tokens")) {
                out.add(runInvalidToken(sharedKey, el.getAsJsonObject()));
            }
        }
        if (doc.has("issue_and_verify_roundtrip")) {
            for (JsonElement el : doc.getAsJsonArray("issue_and_verify_roundtrip")) {
                out.add(runRoundtrip(sharedKey, el.getAsJsonObject()));
            }
        }
        return out;
    }

    private static VectorResult runScopeMatch(JsonObject v) {
        String name = "scope_match/" + v.get("name").getAsString();
        List<String> granted = new ArrayList<>();
        for (JsonElement g : v.getAsJsonArray("granted")) granted.add(g.getAsString());
        String requested = v.get("requested").getAsString();
        boolean expected = v.get("expected").getAsBoolean();
        try {
            boolean got = Scopes.granted(granted, requested);
            if (got != expected) {
                return VectorResult.fail("consent", name, "got " + got + ", want " + expected);
            }
            return VectorResult.pass("consent", name);
        } catch (RuntimeException e) {
            return VectorResult.fail("consent", name, "unexpected exception: " + e.getMessage());
        }
    }

    private static VectorResult runScopeMatchError(JsonObject v) {
        String name = "scope_match_errors/" + v.get("name").getAsString();
        List<String> granted = new ArrayList<>();
        for (JsonElement g : v.getAsJsonArray("granted")) granted.add(g.getAsString());
        String requested = v.get("requested").getAsString();
        String expectErr = v.get("expect_error").getAsString();
        try {
            Scopes.granted(granted, requested);
            return VectorResult.fail("consent", name,
                "expected error containing '" + expectErr + "', got Ok");
        } catch (IllegalArgumentException e) {
            String msg = e.getMessage().toLowerCase();
            if (!msg.contains(expectErr.toLowerCase())) {
                return VectorResult.fail("consent", name,
                    "expected error containing '" + expectErr + "', got: " + e.getMessage());
            }
            return VectorResult.pass("consent", name);
        }
    }

    private static Algorithm parseAlg(JsonObject v) {
        if (v.has("algorithm")) {
            try { return Algorithm.fromWire(v.get("algorithm").getAsString()); }
            catch (Exception ignored) {}
        }
        return Algorithm.HS256;
    }

    private static String stringOrDefault(JsonObject v, String key, String dflt) {
        if (v.has(key) && !v.get(key).isJsonNull()) return v.get(key).getAsString();
        return dflt;
    }

    private static VectorResult runValidToken(String sharedKey, JsonObject v) {
        String name = "valid_tokens/" + v.get("name").getAsString();
        long now = v.get("now").getAsLong();
        String issuer = stringOrDefault(v, "issuer", "iaiso");
        Algorithm alg = parseAlg(v);
        String token = v.get("token").getAsString();
        JsonObject expected = v.getAsJsonObject("expected");

        Verifier verifier = Verifier.builder()
            .hsKey(sharedKey.getBytes())
            .algorithm(alg)
            .issuer(issuer)
            .clock(() -> now)
            .build();
        try {
            Scope s = verifier.verify(token, null);
            if (!s.getSubject().equals(expected.get("sub").getAsString())) {
                return VectorResult.fail("consent", name,
                    "sub: got " + s.getSubject() + ", want " + expected.get("sub"));
            }
            if (!s.getJti().equals(expected.get("jti").getAsString())) {
                return VectorResult.fail("consent", name,
                    "jti: got " + s.getJti() + ", want " + expected.get("jti"));
            }
            List<String> wantScopes = new ArrayList<>();
            for (JsonElement el : expected.getAsJsonArray("scopes")) wantScopes.add(el.getAsString());
            if (!s.getScopes().equals(wantScopes)) {
                return VectorResult.fail("consent", name, "scopes mismatch");
            }
            // execution_id may be null
            String wantExec = expected.has("execution_id") && !expected.get("execution_id").isJsonNull()
                ? expected.get("execution_id").getAsString() : null;
            if (wantExec == null ? s.getExecutionId() != null : !wantExec.equals(s.getExecutionId())) {
                return VectorResult.fail("consent", name,
                    "execution_id: got " + s.getExecutionId() + ", want " + wantExec);
            }
            return VectorResult.pass("consent", name);
        } catch (RuntimeException e) {
            return VectorResult.fail("consent", name, "verify failed: " + e.getMessage());
        }
    }

    private static VectorResult runInvalidToken(String sharedKey, JsonObject v) {
        String name = "invalid_tokens/" + v.get("name").getAsString();
        long now = v.get("now").getAsLong();
        String issuer = stringOrDefault(v, "issuer", "iaiso");
        Algorithm alg = parseAlg(v);
        String token = v.get("token").getAsString();
        String execId = stringOrDefault(v, "execution_id", null);
        String expectErr = v.get("expect_error").getAsString();

        Verifier verifier = Verifier.builder()
            .hsKey(sharedKey.getBytes())
            .algorithm(alg)
            .issuer(issuer)
            .clock(() -> now)
            .build();
        try {
            verifier.verify(token, execId);
            return VectorResult.fail("consent", name,
                "expected error '" + expectErr + "', got Ok");
        } catch (ConsentException.ExpiredToken e) {
            if (!"expired".equals(expectErr)) {
                return VectorResult.fail("consent", name,
                    "expected '" + expectErr + "', got expired");
            }
            return VectorResult.pass("consent", name);
        } catch (ConsentException.RevokedToken e) {
            if (!"revoked".equals(expectErr)) {
                return VectorResult.fail("consent", name,
                    "expected '" + expectErr + "', got revoked");
            }
            return VectorResult.pass("consent", name);
        } catch (ConsentException.InvalidToken e) {
            if (!"invalid".equals(expectErr)) {
                return VectorResult.fail("consent", name,
                    "expected '" + expectErr + "', got invalid: " + e.getMessage());
            }
            return VectorResult.pass("consent", name);
        } catch (RuntimeException e) {
            return VectorResult.fail("consent", name,
                "unexpected exception type: " + e.getClass().getSimpleName() + ": " + e.getMessage());
        }
    }

    private static VectorResult runRoundtrip(String sharedKey, JsonObject v) {
        String name = "roundtrip/" + v.get("name").getAsString();
        JsonObject issueSpec = v.getAsJsonObject("issue");
        long ttl = issueSpec.has("ttl_seconds") ? issueSpec.get("ttl_seconds").getAsLong() : 3600;
        String subject = issueSpec.get("subject").getAsString();
        List<String> scopes = new ArrayList<>();
        for (JsonElement el : issueSpec.getAsJsonArray("scopes")) scopes.add(el.getAsString());
        String execId = stringOrDefault(issueSpec, "execution_id", null);
        JsonObject metadata = issueSpec.has("metadata") && issueSpec.get("metadata").isJsonObject()
            ? issueSpec.getAsJsonObject("metadata") : null;

        long now = v.has("now") && !v.get("now").isJsonNull()
            ? v.get("now").getAsLong() : 1_700_000_000L;
        String issuer = stringOrDefault(v, "issuer", "iaiso");
        Algorithm alg = parseAlg(v);

        Issuer is = Issuer.builder()
            .hsKey(sharedKey.getBytes())
            .algorithm(alg)
            .issuer(issuer)
            .clock(() -> now)
            .build();
        Scope scope;
        try {
            scope = is.issue(subject, scopes, execId, ttl, metadata);
        } catch (RuntimeException e) {
            return VectorResult.fail("consent", name, "issue failed: " + e.getMessage());
        }

        boolean expectSuccess = v.has("expected_after_verify_succeeds")
            && v.get("expected_after_verify_succeeds").getAsBoolean();
        String verifyExec = stringOrDefault(v, "verify_with_execution_id", null);

        Verifier ver = Verifier.builder()
            .hsKey(sharedKey.getBytes())
            .algorithm(alg)
            .issuer(issuer)
            .clock(() -> now + 1)
            .build();
        try {
            Scope verified = ver.verify(scope.getToken(), verifyExec);
            if (!expectSuccess) {
                return VectorResult.fail("consent", name, "expected verify to fail, succeeded");
            }
            if (!verified.getSubject().equals(subject)) {
                return VectorResult.fail("consent", name, "subject mismatch");
            }
            if (!verified.getScopes().equals(scopes)) {
                return VectorResult.fail("consent", name, "scopes mismatch");
            }
            return VectorResult.pass("consent", name);
        } catch (RuntimeException e) {
            if (expectSuccess) {
                return VectorResult.fail("consent", name,
                    "expected verify to succeed, failed: " + e.getMessage());
            }
            return VectorResult.pass("consent", name);
        }
    }
}

File: IAIso-v5.0/core/iaiso-rust/crates/conformance/src/consent_runner.rs
use crate::VectorResult;
use iaiso_consent::{
    scope_granted, ConsentError, Issuer, IssuerOptions, RevocationList, SignatureAlgorithm,
    Verifier, VerifierOptions,
};
use serde::Deserialize;
use std::path::Path;
use std::sync::Arc;

#[derive(Debug, Deserialize)]
struct ScopeMatchVector {
    name: String,
    granted: Vec<String>,
    requested: String,
    expected: bool,
}

#[derive(Debug, Deserialize)]
struct ScopeMatchErrorVector {
    name: String,
    granted: Vec<String>,
    requested: String,
    expect_error: String,
}

#[derive(Debug, Deserialize)]
struct ValidTokenExpected {
    sub: String,
    jti: String,
    scopes: Vec<String>,
    #[serde(default)]
    execution_id: Option<String>,
}

#[derive(Debug, Deserialize)]
struct ValidTokenVector {
    name: String,
    token: String,
    #[serde(default = "default_alg")]
    algorithm: String,
    #[serde(default = "default_iss")]
    issuer: String,
    now: i64,
    expected: ValidTokenExpected,
}

#[derive(Debug, Deserialize)]
struct InvalidTokenVector {
    name: String,
    token: String,
    #[serde(default = "default_alg")]
    algorithm: String,
    #[serde(default = "default_iss")]
    issuer: String,
    now: i64,
    #[serde(default)]
    execution_id: Option<String>,
    expect_error: String,
}

#[derive(Debug, Deserialize)]
struct RoundtripIssue {
    subject: String,
    scopes: Vec<String>,
    ttl_seconds: i64,
    #[serde(default)]
    execution_id: Option<String>,
    #[serde(default)]
    metadata: Option<serde_json::Value>,
}

#[derive(Debug, Deserialize)]
struct RoundtripVector {
    name: String,
    issue: RoundtripIssue,
    #[serde(default)]
    verify_with_execution_id: Option<String>,
    #[serde(default = "default_alg")]
    algorithm: String,
    #[serde(default = "default_iss")]
    issuer: String,
    #[serde(default = "default_now")]
    now: i64,
    #[serde(default)]
    expected_after_verify_succeeds: bool,
}

fn default_now() -> i64 {
    1_700_000_000
}

fn default_alg() -> String {
    "HS256".to_string()
}
fn default_iss() -> String {
    "iaiso".to_string()
}

#[derive(Debug, Deserialize)]
struct ConsentFile {
    hs256_key_shared: String,
    scope_match: Vec<ScopeMatchVector>,
    scope_match_errors: Vec<ScopeMatchErrorVector>,
    valid_tokens: Vec<ValidTokenVector>,
    invalid_tokens: Vec<InvalidTokenVector>,
    issue_and_verify_roundtrip: Vec<RoundtripVector>,
}

fn algorithm_from(s: &str) -> SignatureAlgorithm {
    match s {
        "RS256" => SignatureAlgorithm::RS256,
        _ => SignatureAlgorithm::HS256,
    }
}

pub fn run_consent(spec_root: &Path) -> Result<Vec<VectorResult>, String> {
    let path = spec_root.join("consent").join("vectors.json");
    let data = std::fs::read(&path).map_err(|e| format!("read {}: {}", path.display(), e))?;
    let file: ConsentFile =
        serde_json::from_slice(&data).map_err(|e| format!("parse {}: {}", path.display(), e))?;

    let mut results = Vec::new();
    for v in file.scope_match {
        let mut r = VectorResult {
            section: "consent".to_string(),
            name: format!("scope_match/{}", v.name),
            passed: false,
            message: String::new(),
        };
        match scope_granted(&v.granted, &v.requested) {
            Ok(b) if b == v.expected => r.passed = true,
            Ok(b) => r.message = format!("got {}, want {}", b, v.expected),
            Err(e) => r.message = e.to_string(),
        }
        results.push(r);
    }
    for v in file.scope_match_errors {
        let mut r = VectorResult {
            section: "consent".to_string(),
            name: format!("scope_match_errors/{}", v.name),
            passed: false,
            message: String::new(),
        };
        match scope_granted(&v.granted, &v.requested) {
            Ok(_) => {
                r.message = format!("expected error containing {:?}, got Ok", v.expect_error);
            }
            Err(e) => {
                let msg = e.to_string();
                if !msg.contains(&v.expect_error) {
                    r.message = format!(
                        "expected error containing {:?}, got {:?}",
                        v.expect_error, msg
                    );
                } else {
                    r.passed = true;
                }
            }
        }
        results.push(r);
    }
    for v in file.valid_tokens {
        results.push(run_valid_token(&file.hs256_key_shared, v));
    }
    for v in file.invalid_tokens {
        results.push(run_invalid_token(&file.hs256_key_shared, v));
    }
    for v in file.issue_and_verify_roundtrip {
        results.push(run_roundtrip(&file.hs256_key_shared, v));
    }
    Ok(results)
}

fn run_valid_token(shared_key: &str, v: ValidTokenVector) -> VectorResult {
    let mut r = VectorResult {
        section: "consent".to_string(),
        name: format!("valid_tokens/{}", v.name),
        passed: false,
        message: String::new(),
    };
    let now = v.now;
    let verifier = Verifier::new(VerifierOptions {
        verification_key: shared_key.as_bytes().to_vec(),
        algorithm: algorithm_from(&v.algorithm),
        issuer: v.issuer.clone(),
        revocation_list: None,
        leeway_seconds: 5,
        clock: Some(Arc::new(move || now)),
    });
    match verifier.verify(&v.token, None) {
        Ok(scope) => {
            if scope.subject != v.expected.sub {
                r.message = format!("sub: got {}, want {}", scope.subject, v.expected.sub);
                return r;
            }
            if scope.jti != v.expected.jti {
                r.message = format!("jti: got {}, want {}", scope.jti, v.expected.jti);
                return r;
            }
            if scope.scopes != v.expected.scopes {
                r.message = "scopes mismatch".to_string();
                return r;
            }
            if scope.execution_id != v.expected.execution_id {
                r.message = format!(
                    "execution_id: got {:?}, want {:?}",
                    scope.execution_id, v.expected.execution_id
                );
                return r;
            }
            r.passed = true;
        }
        Err(e) => r.message = format!("verify failed: {}", e),
    }
    r
}

fn run_invalid_token(shared_key: &str, v: InvalidTokenVector) -> VectorResult {
    let mut r = VectorResult {
        section: "consent".to_string(),
        name: format!("invalid_tokens/{}", v.name),
        passed: false,
        message: String::new(),
    };
    let now = v.now;
    let verifier = Verifier::new(VerifierOptions {
        verification_key: shared_key.as_bytes().to_vec(),
        algorithm: algorithm_from(&v.algorithm),
        issuer: v.issuer.clone(),
        revocation_list: None,
        leeway_seconds: 5,
        clock: Some(Arc::new(move || now)),
    });
    match verifier.verify(&v.token, v.execution_id.as_deref()) {
        Ok(_) => r.message = "expected error, got Ok".to_string(),
        Err(e) => {
            let msg = e.to_string().to_lowercase();
            let ok = match v.expect_error.as_str() {
                "invalid" => matches!(e, ConsentError::InvalidToken(_)),
                "expired" => matches!(e, ConsentError::ExpiredToken) || msg.contains("expir"),
                "revoked" => matches!(e, ConsentError::RevokedToken(_)) || msg.contains("revok"),
                other => msg.contains(other),
            };
            if ok {
                r.passed = true;
            } else {
                r.message = format!("expected error of kind {}, got: {}", v.expect_error, e);
            }
        }
    }
    r
}

fn run_roundtrip(shared_key: &str, v: RoundtripVector) -> VectorResult {
    let mut r = VectorResult {
        section: "consent".to_string(),
        name: format!("roundtrip/{}", v.name),
        passed: false,
        message: String::new(),
    };
    let algo = algorithm_from(&v.algorithm);
    let now = v.now;

    let issuer = Issuer::new(IssuerOptions {
        signing_key: shared_key.as_bytes().to_vec(),
        algorithm: algo,
        issuer: v.issuer.clone(),
        default_ttl_seconds: 3600,
        clock: Some(Arc::new(move || now)),
    });
    let scope = match issuer.issue(
        &v.issue.subject,
        v.issue.scopes.clone(),
        v.issue.execution_id.clone(),
        Some(v.issue.ttl_seconds),
        v.issue.metadata.clone(),
    ) {
        Ok(s) => s,
        Err(e) => {
            r.message = format!("issue failed: {}", e);
            return r;
        }
    };

    let now_plus = v.now + 1;
    let verifier = Verifier::new(VerifierOptions {
        verification_key: shared_key.as_bytes().to_vec(),
        algorithm: algo,
        issuer: v.issuer.clone(),
        revocation_list: None,
        leeway_seconds: 5,
        clock: Some(Arc::new(move || now_plus)),
    });
    let verify_result = verifier.verify(&scope.token, v.verify_with_execution_id.as_deref());
    if v.expected_after_verify_succeeds {
        match verify_result {
            Ok(verified) => {
                if verified.subject != v.issue.subject {
                    r.message = format!(
                        "subject mismatch: {} vs {}",
                        verified.subject, v.issue.subject
                    );
                    return r;
                }
                if verified.scopes != v.issue.scopes {
                    r.message = "scopes mismatch".to_string();
                    return r;
                }
                let _ = RevocationList::new(); // exercise the type
                r.passed = true;
            }
            Err(e) => {
                r.message = format!("expected verify to succeed, failed: {}", e);
            }
        }
    } else if verify_result.is_err() {
        r.passed = true;
    } else {
        r.message = "expected verify to fail, succeeded".to_string();
    }
    r
}
